src/Mvc/Mvc.ViewFeatures/src/Infrastructure/DefaultTempDataSerializer.cs

```diff
@@ -3,30 +3,142 @@
 
 using System;
 using System.Collections.Generic;
-using Microsoft.Extensions.DependencyInjection;
+using System.Globalization;
+using System.Text.Json;
 
 namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Infrastructure
 {
     internal class DefaultTempDataSerializer : TempDataSerializer
     {
-        public override IDictionary<string, object> Deserialize(byte[] unprotectedData)
+        public override IDictionary<string, object> Deserialize(byte[] value)
         {
-            throw new InvalidOperationException(Core.Resources.FormatReferenceToNewtonsoftJsonRequired(
-                Resources.DeserializingTempData,
-                "Microsoft.AspNetCore.Mvc.NewtonsoftJson",
-                nameof(IMvcBuilder),
-                "AddNewtonsoftJson",
-                "ConfigureServices(...)"));
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            if (value.Length == 0)
+            {
+                return new Dictionary<string, object>();
+            }
+
+            using var jsonDocument = JsonDocument.Parse(value);
+            var rootElement = jsonDocument.RootElement;
+            return DeserializeDictionary(rootElement);
+        }
+
+        private IDictionary<string, object> DeserializeDictionary(JsonElement rootElement)
+        {
+            var deserialized = new Dictionary<string, object>(StringComparer.Ordinal);
+
+            foreach (var item in rootElement.EnumerateObject())
+            {
+                object deserializedValue;
+                switch (item.Value.Type)
+                {
+                    case JsonValueType.False:
+                    case JsonValueType.True:
+                        deserializedValue = item.Value.GetBoolean();
+                        break;
+
+                    case JsonValueType.Number:
+                        deserializedValue = item.Value.GetInt32();
+                        break;
+
+                    case JsonValueType.String:
+                        var stringValue = item.Value.GetString();
+                        if (DateTime.TryParseExact(stringValue, "o", CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out var dateTime))
```

Comment:
Do we keep track of the types we're putting in? 


Seeing logic like this just seems totally flawed? What if a string happened to contain a valid date. I think we have to roundtrip the type names. What did we do with the JSON.NET version?

Reply:
With Json.NET, `TypeNameHandling` is null, so we get Json.NET's best effort behavior. From what I can tell,

1) `string`, `bools` are always round-tripped correctly.
2) `int` get round-tripped as `long`. We do special handling for this. Consequently things like `short`, `byte`, or a `long` that fits inside an `int` don't round-trip
3) Json.NET returns a `DateTime` for any string that can be parsed by `DateTime.Parse`. Same goes for `Guid`. However, `DateTimeOffset` round trip as a `DateTime` instance, `TimeSpan` and `Uri` as a `string` value

We could in theory enable `TypeNameHandling` in Json.NET's implementation to get better behavior, part of why I filed #8873, but that flag's controversial. 

For this specific bit, we could remove the special support for `DateTime`. We have one test that fails as a result, but I can change it to do a different data type

Reply:
Ok I didn't realize that's what we did. I think we should try to make this highly compatible with newtonsoft if possible.